### DISCUSSION TITLE
❓ Phase 23.4 Q&A — DecisionTreeSolver: MCTS tuning, alpha-beta pruning & game-theoretic reasoning

### DISCUSSION THREAD

**Discussion by web3guru888** [web3guru888/asi-build#542] [Q&A]
# ❓ Phase 23.4 Q&A — DecisionTreeSolver

Common questions about the **DecisionTreeSolver** component (Phase 23.4, Issue #532).

---

## Q1: When should the agent use MCTS vs exact backward induction?

**A:** It depends on tree size and time budget:

| Criterion | Backward Induction | MCTS |
|-----------|-------------------|------|
| Tree size | Small (< ~10⁶ nodes) | Large / unknown |
| Guarantee | Provably optimal | Converges to optimal |
| Model needed? | Full tree enumerable | Only need simulator |
| Anytime? | No — must finish | Yes — returns best-so-far |
| Adversarial? | Only if minimax variant | Handles via UCT |

**Rule of thumb**: if you can enumerate the full tree in < 100ms, use backward induction for its optimality guarantee. For anything larger — especially when the branching factor exceeds ~20 or depth exceeds ~15 — MCTS gives good results within a time budget. The `solve()` method's `algorithm` parameter lets you override the automatic selection.

---

## Q2: Why is the UCB1 exploration constant c = √2?

**A:** The value c = √2 is the **theoretically optimal** exploration constant for rewards bounded in [0, 1], derived from the Hoeffding inequality applied to the multi-armed bandit setting (Auer et al., 2002). It balances:

- **Exploitation** (Q̄ term) — prefer nodes with high average reward
- **Exploration** (√(ln N_parent / N_node) term) — prefer under-visited nodes

In practice, the optimal c depends on the reward scale and domain structure:

```python
# Default in DecisionTreeConfig
exploration_constant: float = 1.414  # √2

# For risk-averse domains (tighter exploitation)
exploration_constant: float = 0.5

# For highly stochastic domains (more exploration)
exploration_constant: float = 2.0
```

**Recommendation**: Start with √2, then tune on a validation set of decision problems. Values between 0.5 and 2.5 cover most practical domains.

---

## Q3: How does alpha-beta pruning help, and when does it fail?

**A:** Alpha-beta pruning eliminates branches that **provably cannot affect** the minimax value:

- **Alpha** (α) = best value MAX can guarantee so far
- **Beta** (β) = best value MIN can guarantee so far
- When α ≥ β at any node, we **cut** the remaining children

**Complexity reduction**:
```
Without pruning:  O(b^d)        — examine every leaf
With pruning:     O(b^(d/2))    — best case (perfect move ordering)
Average case:     O(b^(3d/4))   — random move ordering
```

**When it fails** (no pruning benefit):
- Worst-case move ordering (adversary's best moves evaluated last)
- Non-zero-sum games (minimax assumption violated)
- Stochastic nodes between MIN/MAX layers (use expectimax instead)

**Mitigation**: The solver uses **iterative deepening** — solving depth d-1 first to get move ordering hints for depth d. This typically achieves near-best-case pruning.

---

## Q4: Value of Information — when is it worth computing?

**A:** Computing VoI requires solving the tree **twice** (uninformed and informed), so it's expensive. Use it when:

```
VoI_decision_rule:
    IF decision_stakes > voi_threshold:          # high-consequence decision
        IF unresolved_chance_nodes exist:         # uncertainty present
            IF information_cost is known:         # can we actually acquire info?
                compute VoI
                IF VoI > information_cost:
                    acquire information first
                ELSE:
                    decide with current beliefs
```

**Practical heuristics**:
- Skip VoI for trees with < 3 CHANCE nodes (overhead not worth it)
- Cache partial VoI results — if resolving node A doesn't change the optimal action, skip nodes downstream of A
- Use **Expected Value of Perfect Information (EVPI)** as an upper bound — if EVPI < cost, no individual VoI can exceed cost either

---

## Q5: How does the solver handle continuous action spaces?

**A:** Pure tree search requires discrete actions, but real-world action spaces are often continuous. Three strategies:

**1. Discretisation** (simplest):
```python
# Uniform grid over action range
actions = np.linspace(a_min, a_max, num_bins)
```

**2. Progressive Widening** (MCTS-specific):
```python
# Add new child only when visit count exceeds threshold
if len(node.children) < ceil(node.visits ** alpha):
    new_action = sample_from_prior(node.state)
    node.add_child(new_action)
# alpha ∈ (0, 1) controls widening rate
# alpha = 0.5 is typical (√N children after N visits)
```

**3. Continuous MCTS extensions** (advanced):
- **HOLP** (Hierarchical Optimistic Planning) for Lipschitz-continuous rewards
- **CEM-MCTS** — use Cross-Entropy Method to sample promising actions

The `DecisionTreeConfig.widening_alpha` parameter controls progressive widening. Set to `None` to disable (finite action spaces only).

---

## Q6: How does asyncio compatibility work for large trees?

**A:** Tree search can run for millions of nodes. To avoid blocking the event loop:

```python
async def solve(self, root: DecisionNode) -> SolveResult:
    nodes_expanded = 0
    async for action in self._search(root):
        nodes_expanded += 1
        if nodes_expanded % self._config.yield_every_n == 0:
            await asyncio.sleep(0)  # yield to event loop
    return self._best_result(root)
```

**Configuration**:
```python
@dataclasses.dataclass(frozen=True)
class DecisionTreeConfig:
    yield_every_n: int = 1000      # yield control every N expansions
    max_wall_seconds: float = 5.0  # hard time limit
    max_iterations: int = 100_000  # hard iteration limit
```

- `yield_every_n = 1000` means the solver yields control to asyncio after every 1000 node expansions
- Combined with `max_wall_seconds`, this ensures the CognitiveCycle is never starved
- For MCTS specifically, each iteration is lightweight (one path), so yielding every 100 iterations works well

---

## Q7: How does DecisionTreeSolver integrate with WorldModel (Phase 13)?

**A:** WorldModel provides the **state transition function** that MCTS needs for simulation:

```python
class WorldModel(Protocol):
    async def predict_transition(
        self, state: State, action: Action
    ) -> tuple[State, float]:  # (next_state, reward)
        ...

class AsyncDecisionTreeSolver:
    def __init__(
        self,
        config: DecisionTreeConfig,
        world_model: WorldModel,        # Phase 13
        utility: UtilityComputer,       # Phase 23.3
        uncertainty: UncertaintyQuantifier,  # Phase 23.1
    ):
        ...
```

**Integration points**:

| MCTS Phase | WorldModel Usage |
|------------|------------------|
| **Expansion** | `predict_transition(state, action)` → child state |
| **Simulation** | Repeated `predict_transition` for rollout trajectory |
| **Terminal check** | `world_model.is_terminal(state)` → stop rollout |

**Without WorldModel** (fallback): If no WorldModel is injected, the solver operates in **explicit tree mode** — the caller must provide the full `DecisionNode` tree upfront. This is useful for combinatorial problems (scheduling, resource allocation) where the tree is pre-built.

---

*See also*:
- Issue #532 — DecisionTreeSolver full spec
- Discussion — [Show & Tell 23.4](../discussions) for architecture diagrams
- Phase 23 Planning → #533
- UncertaintyQuantifier 23.1 → #529
- UtilityComputer 23.3 → #531
